sdk/tests/conformance/extensions/khr-parallel-shader-compile.html

```diff
@@ -58,36 +95,131 @@
+
+        let start = performance.now();
+        gl.getShaderParameter(fs, COMPLETION_STATUS_KHR);
+        gl.getShaderParameter(vs, COMPLETION_STATUS_KHR);
+        let duration = performance.now() - start;
+        if (performance.now() - start > 100)
+            testFailed(`Querying shader status should not wait for compilation. Took ${duration} ms`);
+
+        let frames = 0;
+        while (!gl.getProgramParameter(program, COMPLETION_STATUS_KHR)
+            && performance.now() - start < measuredCompileDuration * 4) {
+            frames++;
+            await new Promise(requestAnimationFrame);
+        }
+        duration = performance.now() - start;
+        if (!gl.getProgramParameter(program, COMPLETION_STATUS_KHR)) {
+            testFailed(`Program took longer than 4x the expected duration to compile. Expected: ${measuredCompileDuration} ms, actual: ${duration} ms`);
+        } else if (!gl.getShaderParameter(vs, COMPLETION_STATUS_KHR) || !gl.getShaderParameter(fs, COMPLETION_STATUS_KHR)) {
+            testFailed('Program linked before shaders finished compiling.');
+        } else if (frames <= 6) {
+            testFailed(`Program should have taken many more than 6 frames to compile. Actual value: ${frames} frames, duration ${performance.now() - start} ms.`);
+        } else {
+            console.log(`COMPLETION_STATUS_KHR sucessfully transitioned from false to true in ${frames} frames and ${duration} ms.`);
+            testPassed(`COMPLETION_STATUS_KHR sucessfully transitioned from false to true`);
+        }
+
+
+        debug("Checking that status is true when context is lost.");
+        if (loseContext) {
+            gl.shaderSource(vs, vertexSource(extraCode));
+            gl.shaderSource(fs, fragmentSource(extraCode));
+            gl.compileShader(vs);
+            gl.compileShader(fs);
+            gl.linkProgram(program);
+            loseContext.loseContext();
+            status = gl.getShaderParameter(vs, COMPLETION_STATUS_KHR);
+            if (status !== true) testFailed("shader status should be true when context is lost");
+            status = gl.getProgramParameter(program, COMPLETION_STATUS_KHR);
+            if (status !== true) testFailed("program status should be true when context is lost");
+            loseContext.restoreContext();
+            vs = gl.createShader(gl.VERTEX_SHADER);
+            fs = gl.createShader(gl.FRAGMENT_SHADER);
+            program = gl.createProgram();
+        }
+    }
+    finishTest();
+})();
+
+async function compileShaders() {
+    console.log('Compiling shaders');
+    const gl = canvas.getContext('webgl');
+    const vs = gl.createShader(gl.VERTEX_SHADER);
+    const fs = gl.createShader(gl.FRAGMENT_SHADER);
+    const program = gl.createProgram();
+    gl.getExtension(wtu.getExtensionWithKnownPrefixes(gl, "KHR_parallel_shader_compile"));
+    gl.attachShader(program, vs);
+    gl.attachShader(program, fs);
+    gl.shaderSource(vs, vertexSource(extraCode));
```

Comment:
Since invoking this function doesn't change extraCode, it seems likely to me that these shader compilations and program link could hit the browser's program binary caches, and not actually do any work. Would it be possible to better randomize the shader's code? Repeatability is ideal, and a pseudorandom number generator would work for that (see http://crbug.com/1132587#c9 ), but that would also probably hit the browser's cache from run to run. Maybe seeding the PRNG with the current time and printing the seed for each run would allow debugging and preserve repeatability.

Reply:
The vertexSource and fragmentSource functions increment a counter and put the changed value into the code every time they are called, so it should bypass any cache.

Reply:
I see - thanks for pointing that out.